cfg/optimize.go

```
@@ -0,0 +1,294 @@
+		if machineType != "" {
+			return machineType, nil
+		}
+	}
+	client := http.Client{Timeout: httpTimeout}
+	for retry := 0; retry < maxRetries; retry++ {
+		for _, endpoint := range metadataEndpoints {
+
+			req, err := http.NewRequest(http.MethodGet, endpoint, nil)
+			if err != nil {
+				return "", fmt.Errorf("failed to create request for %s: %w", endpoint, err)
+			}
+			req.Header.Add("Metadata-Flavor", "Google")
+
+			resp, err := client.Do(req)
+			if err != nil {
+				continue // Try the next endpoint.
+			}
+			defer resp.Body.Close()
+
+			if resp.StatusCode != http.StatusOK {
+				continue
+			}
+
+			body, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return "", fmt.Errorf("failed to read response body from %s: %w", endpoint, err)
+			}
+
+			machineType := string(body)
+			parts := strings.Split(machineType, "/")
+			return parts[len(parts)-1], nil
+		}
+	}
+
+	return "", fmt.Errorf("failed to get machine type from any metadata endpoint after retries")
+}
+
+// applyMachineTypeOptimizations applies optimizations based on the detected machine type.
+func applyMachineTypeOptimizations(config *optimizationConfig, cfg *Config, isSet isValueSet) ([]string, error) {
+	machineType, err := getMachineType(isSet)
+	if err != nil {
+		return []string{}, nil // Non-fatal error, continue with default settings.
+	}
+	var optimizedFlags []string
+	for _, mt := range config.machineTypes {
+		for _, name := range mt.names {
+			if strings.HasPrefix(machineType, name) {
+				// Find the FlagOverrideSet.
+				var flgOvrrideSet *flagOverrideSet
+				for _, flgOverrideSet := range config.flagOverrideSets {
+					if flgOverrideSet.name == mt.flagOverrideSetName {
+						flgOvrrideSet = &flgOverrideSet
+						break
+					}
+				}
+
+				if flgOvrrideSet == nil {
+					continue
+				}
+
+				for flag, override := range flgOvrrideSet.overrides {
+					// Use reflection to find the field in ServerConfig.
+					err := setFlagValue(cfg, flag, override, isSet)
```

if error is not nil, should we return an error?

---

The reason we are not returning is we want to still do best effort in applying other defaults. Ideally, there should not be any error.

---

We should at least log it then.

---

Indeed but the logger module right now has dependency on cfg module and hence not usable here due to circular reference. Once we break that dependency we can start using it for logging here (printing on stderr does not seem clean), that change would be out of the purview of this.